File: examples/app-baseline/golang/riffdb.go
// RiffDB load session: generated TicketDesk client only. No Go safety code.
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"reflect"
	"strconv"

	application "riffdb.dev/application"
	"riffdb.dev/ticketdesk"
)

const riffdbBackendID = "riffdb_public_grpc"

type loadKindError struct {
	kind string
	msg  string
}

func (e *loadKindError) Error() string { return e.msg }

func classifyRiffdb(err error) string {
	var app *application.ApplicationError
	if errors.As(err, &app) {
		switch app.Details.Code {
		case "RDB-COMMAND-0102":
			return "conflict"
		case "RDB-STORAGE-0101", "RDB-RESOURCE-0101", "RDB-CAPACITY-0101":
			return "unavailable"
		default:
			return "error"
		}
	}
	var kind *loadKindError
	if errors.As(err, &kind) {
		if kind.kind == "conflict" {
			return "conflict"
		}
		return "error"
	}
	return "error"
}

type driverIdentityFile struct {
	SocketPath string
	Identity   application.Identity
}

func readDriverIdentity(path string) (driverIdentityFile, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return driverIdentityFile{}, err
	}
	var parsed struct {
		SocketPath string `json:"socketPath"`
		Identity   struct {
			ApplicationManifestHash string          `json:"applicationManifestHash"`
			OperationCatalogHash    string          `json:"operationCatalogHash"`
			ContractLineage         string          `json:"contractLineage"`
			ContractVersion         json.RawMessage `json:"contractVersion"`
			ContractBundleHash      string          `json:"contractBundleHash"`
			Database                string          `json:"database"`
			Role                    string          `json:"role"`
			RoleDefinitionHash      string          `json:"roleDefinitionHash"`
			RemoteIdentityHash      string          `json:"remoteIdentityHash"`
		} `json:"identity"`
	}
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return driverIdentityFile{}, err
	}
	version, err := parseFlexUint64(parsed.Identity.ContractVersion)
	if err != nil {
		return driverIdentityFile{}, err
	}
	return driverIdentityFile{
		SocketPath: parsed.SocketPath,
		Identity: application.Identity{
			ApplicationManifestHash: parsed.Identity.ApplicationManifestHash,
			OperationCatalogHash:    parsed.Identity.OperationCatalogHash,
			ContractLineage:         parsed.Identity.ContractLineage,
			ContractVersion:         version,
			ContractBundleHash:      parsed.Identity.ContractBundleHash,
			Database:                parsed.Identity.Database,
			Role:                    parsed.Identity.Role,
			RoleDefinitionHash:      parsed.Identity.RoleDefinitionHash,
			RemoteIdentityHash:      parsed.Identity.RemoteIdentityHash,
		},
	}, nil
}

func parseFlexUint64(raw json.RawMessage) (uint64, error) {
	if len(raw) == 0 {
		return 0, errors.New("missing contractVersion")
	}
	if raw[0] == '"' {
		var text string
		if err := json.Unmarshal(raw, &text); err != nil {
			return 0, err
		}
		return strconv.ParseUint(text, 10, 64)
	}
	var value uint64
	if err := json.Unmarshal(raw, &value); err != nil {
		return 0, err
	}
	return value, nil
}

func namedOutcome(value any) string {
	rv := reflect.ValueOf(value)
	if rv.Kind() == reflect.Interface {
		rv = rv.Elem()
	}
	if rv.Kind() == reflect.Struct {
		field := rv.FieldByName("Outcome")
		if field.IsValid() && field.Kind() == reflect.String {
			return field.String()
		}
	}
	return fmt.Sprintf("%T", value)
}

func requireOutcome(actual, expected string) error {
	if actual == expected {
		return nil
	}
	switch actual {
	case "CommentExists", "TicketExists", "LinkExists":
		return &loadKindError{kind: "conflict", msg: fmt.Sprintf("unexpected outcome %s, wanted %s", actual, expected)}
	default:
		return &loadKindError{kind: "error", msg: fmt.Sprintf("unexpected outcome %s, wanted %s", actual, expected)}
	}
}

func u32(value uint32) *uint32 { return &value }

type riffDbSession struct {
	session *application.Session
	client  *ticketdesk.Client
}

func connectRiffdb(ctx context.Context, identity driverIdentityFile) (*riffDbSession, error) {
	session, err := application.Connect(ctx, identity.SocketPath, identity.Identity)
	if err != nil {
		return nil, err
	}
	client, err := ticketdesk.NewClient(session, 1)
	if err != nil {
		_ = session.Close()
		return nil, err
	}
	return &riffDbSession{session: session, client: client}, nil
}

func (s *riffDbSession) Close(ctx context.Context) error {
	_ = ctx
	return s.session.Close()
}

func (s *riffDbSession) Prewarm(ctx context.Context, organizationID, ticketID UUID) error {
	_, err := s.PointGetTicket(ctx, organizationID, ticketID)
	return err
}

func (s *riffDbSession) PointGetTicket(ctx context.Context, organizationID, ticketID UUID) (bool, error) {
	result, err := s.client.GetTicket(ctx, ticketdesk.GetTicketParams{
		OrganizationId: formatUUID(organizationID),
		TicketId:       formatUUID(ticketID),
	}, ticketdesk.QueryOptions{})
	if err != nil {
		return false, err
	}
	_, found := result.Value.(ticketdesk.GetTicketFound)
	return found, nil
}

func (s *riffDbSession) PointGetUser(ctx context.Context, organizationID, userID UUID) (bool, error) {
	result, err := s.client.GetUser(ctx, ticketdesk.GetUserParams{
		OrganizationId: formatUUID(organizationID),
		UserId:         formatUUID(userID),
	}, ticketdesk.QueryOptions{})
	if err != nil {
		return false, err
	}
	_, found := result.Value.(ticketdesk.GetUserFound)
	return found, nil
}

func (s *riffDbSession) ListTicketsByProjectStatus(ctx context.Context, organizationID, projectID UUID, status string, limit int) error {
	_, err := s.client.ListTickets(ctx, ticketdesk.ListTicketsParams{
		OrganizationId: formatUUID(organizationID),
		ProjectId:      formatUUID(projectID),
		Statuses:       []ticketdesk.TicketStatus{ticketdesk.TicketStatus(sqlStatusToRiff(status))},
		Limit:          u32(uint32(limit)),
	}, ticketdesk.QueryOptions{})
	return err
}

func (s *riffDbSession) ListOpenTicketsForAssignee(ctx context.Context, organizationID, assigneeID UUID, limit int) error {
	_, err := s.client.ListTicketsByAssignee(ctx, ticketdesk.ListTicketsByAssigneeParams{
		OrganizationId: formatUUID(organizationID),
		AssigneeId:     formatUUID(assigneeID),
		Statuses:       []ticketdesk.TicketStatus{ticketdesk.TicketStatusOpen},
		Limit:          u32(uint32(limit)),
	}, ticketdesk.QueryOptions{})
	return err
}

func (s *riffDbSession) ListCommentsForTicket(ctx context.Context, organizationID, ticketID UUID, limit int) error {
	_, err := s.client.ListComments(ctx, ticketdesk.ListCommentsParams{
		OrganizationId: formatUUID(organizationID),
		TicketId:       formatUUID(ticketID),
		Limit:          u32(uint32(limit)),
	}, ticketdesk.QueryOptions{})
	return err
}

func (s *riffDbSession) ListProjectMembers(ctx context.Context, organizationID, projectID UUID, _ int) error {
	_, err := s.client.ProjectMembers(ctx, ticketdesk.ProjectMembersParams{
		OrganizationId: formatUUID(organizationID),
		ProjectId:      formatUUID(projectID),
	}, ticketdesk.QueryOptions{})
	return err
}

func (s *riffDbSession) TicketDetailPage(ctx context.Context, organizationID, ticketID UUID) (bool, error) {
	result, err := s.client.TicketPage(ctx, ticketdesk.TicketPageParams{
		OrganizationId: formatUUID(organizationID),
		TicketId:       formatUUID(ticketID),
	}, ticketdesk.QueryOptions{})
	if err != nil {
		return false, err
	}
	_, found := result.Value.(ticketdesk.TicketPageFound)
	return found, nil
}

func (s *riffDbSession) CreateComment(ctx context.Context, comment commentSeed) error {
	result, err := s.client.CreateComment(ctx, ticketdesk.CreateCommentInput{
		Body:           comment.row.body,
		AuthorId:       formatUUID(comment.row.authorID),
		TicketId:       formatUUID(comment.row.ticketID),
		CommentId:      formatUUID(comment.row.commentID),
		IdempotencyKey: comment.idempotencyKey,
		OrganizationId: formatUUID(comment.row.organizationID),
	})
	if err != nil {
		return err
	}
	return requireOutcome(namedOutcome(result.Outcome), "Created")
}

func (s *riffDbSession) CloseTicketWithComment(ctx context.Context, input closeTicketWithCommentSeed) error {
	result, err := s.client.CloseTicketWithComment(ctx, ticketdesk.CloseTicketWithCommentInput{
		Body:           input.body,
		AuthorId:       formatUUID(input.authorID),
		TicketId:       formatUUID(input.ticketID),
		CommentId:      formatUUID(input.commentID),
		IdempotencyKey: input.idempotencyKey,
		OrganizationId: formatUUID(input.organizationID),
	})
	if err != nil {
		return err
	}
	return requireOutcome(namedOutcome(result.Outcome), "Closed")
}

func (s *riffDbSession) OpenTicketWithLabels(ctx context.Context, input openTicketWithLabelsSeed) error {
	result, err := s.client.OpenTicketWithLabels(ctx, ticketdesk.OpenTicketWithLabelsInput{
		Title:          input.title,
		LabelA:         formatUUID(input.labelA),
		LabelB:         formatUUID(input.labelB),
		TicketId:       formatUUID(input.ticketID),
		ProjectId:      formatUUID(input.projectID),
		AssigneeId:     formatUUID(input.assigneeID),
		ReporterId:     formatUUID(input.reporterID),
		IdempotencyKey: input.idempotencyKey,
		OrganizationId: formatUUID(input.organizationID),
	})
	if err != nil {
		return err
	}
	return requireOutcome(namedOutcome(result.Outcome), "Created")
}

func (s *riffDbSession) Seed(ctx context.Context, dataset seedDataset) error {
	for _, org := range dataset.organizations {
		result, err := s.client.CreateOrganization(ctx, ticketdesk.CreateOrganizationInput{
			Name:           org.name,
			OrganizationId: formatUUID(org.id),
			IdempotencyKey: "seed-org-" + encodeShort(org.id),
		})
		if err != nil {
			return err
		}
		if err := requireOutcome(namedOutcome(result.Outcome), "Created"); err != nil {
			return err
		}
	}
	for _, user := range dataset.users {
		result, err := s.client.CreateUser(ctx, ticketdesk.CreateUserInput{
			Email:          user.email,
			UserId:         formatUUID(user.userID),
			DisplayName:    user.display,
			OrganizationId: formatUUID(user.organizationID),
			IdempotencyKey: "seed-user-" + encodeShort(user.userID),
		})
		if err != nil {
			return err
		}
		if err := requireOutcome(namedOutcome(result.Outcome), "Created"); err != nil {
			return err
		}
	}
	for _, project := range dataset.projects {
		result, err := s.client.CreateProject(ctx, ticketdesk.CreateProjectInput{
			Name:           project.name,
			ProjectId:      formatUUID(project.projectID),
			OrganizationId: formatUUID(project.organizationID),
			IdempotencyKey: "seed-project-" + encodeShort(project.projectID),
		})
		if err != nil {
			return err
		}
		if err := requireOutcome(namedOutcome(result.Outcome), "Created"); err != nil {
			return err
		}
	}
	for _, member := range dataset.members {
		result, err := s.client.AddProjectMember(ctx, ticketdesk.AddProjectMemberInput{
			Role:           member.role,
			UserId:         formatUUID(member.userID),
			ProjectId:      formatUUID(member.projectID),
			OrganizationId: formatUUID(member.organizationID),
			IdempotencyKey: "seed-member-" + encodeShort(member.projectID) + "-" + encodeShort(member.userID),
		})
		if err != nil {
			return err
		}
		if err := requireOutcome(namedOutcome(result.Outcome), "Created"); err != nil {
			return err
		}
	}
	for _, label := range dataset.labels {
		result, err := s.client.CreateLabel(ctx, ticketdesk.CreateLabelInput{
			Name:           label.name,
			LabelId:        formatUUID(label.labelID),
			OrganizationId: formatUUID(label.organizationID),
			IdempotencyKey: "seed-label-" + encodeShort(label.labelID),
		})
		if err != nil {
			return err
		}
		if err := requireOutcome(namedOutcome(result.Outcome), "Created"); err != nil {
			return err
		}
	}
	for _, ticket := range dataset.tickets {
		result, err := s.client.CreateTicket(ctx, ticketdesk.CreateTicketInput{
			Title:          ticket.title,
			Status:         ticketdesk.TicketStatus(sqlStatusToRiff(ticket.status)),
			TicketId:       formatUUID(ticket.ticketID),
			ProjectId:      formatUUID(ticket.projectID),
			AssigneeId:     formatUUID(ticket.assigneeID),
			ReporterId:     formatUUID(ticket.reporterID),
			OrganizationId: formatUUID(ticket.organizationID),
			IdempotencyKey: "seed-ticket-" + encodeShort(ticket.ticketID),
		})
		if err != nil {
			return err
		}
		if err := requireOutcome(namedOutcome(result.Outcome), "Created"); err != nil {
			return err
		}
	}
	for _, comment := range dataset.comments {
		result, err := s.client.CreateComment(ctx, ticketdesk.CreateCommentInput{
			Body:           comment.body,
			AuthorId:       formatUUID(comment.authorID),
			TicketId:       formatUUID(comment.ticketID),
			CommentId:      formatUUID(comment.commentID),
			OrganizationId: formatUUID(comment.organizationID),
			IdempotencyKey: "seed-comment-" + encodeShort(comment.commentID),
		})
		if err != nil {
			return err
		}
		if err := requireOutcome(namedOutcome(result.Outcome), "Created"); err != nil {
			return err
		}
	}
	for _, link := range dataset.ticketLabels {
		result, err := s.client.AttachLabel(ctx, ticketdesk.AttachLabelInput{
			LabelId:        formatUUID(link.labelID),
			TicketId:       formatUUID(link.ticketID),
			OrganizationId: formatUUID(link.organizationID),
			IdempotencyKey: "seed-link-" + encodeShort(link.ticketID) + "-" + encodeShort(link.labelID),
		})
		if err != nil {
			return err
		}
		if err := requireOutcome(namedOutcome(result.Outcome), "Created"); err != nil {
			return err
		}
	}
	return nil
}

type riffDbDriver struct {
	identity driverIdentityFile
}

func (d riffDbDriver) BackendID() string { return riffdbBackendID }

func (d riffDbDriver) Seed(ctx context.Context, dataset seedDataset) error {
	session, err := connectRiffdb(ctx, d.identity)
	if err != nil {
		return err
	}
	defer session.Close(ctx)
	return session.Seed(ctx, dataset)
}

func (d riffDbDriver) OpenSession(ctx context.Context) (loadSession, error) {
	return connectRiffdb(ctx, d.identity)
}

File: examples/app-baseline/typescript/src/riffdb.ts
/** RiffDB load session: generated TicketDesk client only. No TypeScript safety code. */

import { readFile } from "node:fs/promises";

import {
  DriverApplicationError,
  DriverApplicationTransport,
  DriverGeneratedApplicationTransport,
  type DriverApplicationIdentity,
} from "@riffdb/application";

import { encodeShort, formatUuid, sqlStatusToRiff, type UuidBytes } from "./ids.js";
import type {
  CloseTicketWithCommentSeed,
  CommentSeed,
  OpenTicketWithLabelsSeed,
  SeedDataset,
} from "./seed.js";
import * as generatedClient from "./ticketdesk-client.js";

export /** Bounded batch fan-out for seeding, matching the Rust harness. */
const SEED_CONCURRENCY = 128;
/** Generated batch row ceiling. */
const SEED_MAX_BATCH_ROWS = 4096;
/**
 * The driver frames a whole batch as one message bounded at 1 MiB, so the
 * generated row ceiling is not the binding constraint: 4,096 comment inputs
 * carrying 256-byte bodies encode to roughly 1.8 MB and the driver refuses the
 * frame. Target half the budget so per-row variation and framing overhead have
 * headroom, and size each phase from its own inputs rather than guessing one
 * number for eight differently shaped phases.
 */
const SEED_FRAME_TARGET_BYTES = 512 * 1024;

const RIFFDB_BACKEND_ID = "riffdb_public_grpc";

type TicketDeskModule = typeof generatedClient;

async function loadGenerated(): Promise<TicketDeskModule> {
  return generatedClient;
}

export function classifyRiffdb(error: unknown): string {
  if (error instanceof DriverApplicationError) {
    const code = error.details.code;
    if (code === "RDB-COMMAND-0102") return "conflict";
    if (code === "RDB-STORAGE-0101" || code === "RDB-RESOURCE-0101" || code === "RDB-CAPACITY-0101") {
      return "unavailable";
    }
    return "error";
  }
  if (error instanceof RiffDbLoadError) {
    if (error.kind === "conflict") return "conflict";
    return "error";
  }
  return "error";
}

export class RiffDbLoadError extends Error {
  override readonly name = "RiffDbLoadError";
  constructor(
    message: string,
    readonly kind: "conflict" | "error" = "error",
  ) {
    super(message);
  }
}

export interface DriverIdentityFile {
  readonly socketPath: string;
  readonly identity: DriverApplicationIdentity;
}

export async function readDriverIdentity(path: string): Promise<DriverIdentityFile> {
  const parsed = JSON.parse(await readFile(path, "utf8")) as {
    socketPath: string;
    identity: Omit<DriverApplicationIdentity, "contractVersion"> & { contractVersion: string | number };
  };
  return {
    socketPath: parsed.socketPath,
    identity: {
      ...parsed.identity,
      contractVersion: BigInt(parsed.identity.contractVersion),
    },
  };
}

function requireOutcome(actual: string, expected: string): void {
  if (actual === expected) return;
  if (actual === "CommentExists" || actual === "TicketExists" || actual === "LinkExists") {
    throw new RiffDbLoadError(`unexpected outcome ${actual}, wanted ${expected}`, "conflict");
  }
  throw new RiffDbLoadError(`unexpected outcome ${actual}, wanted ${expected}`);
}

export class RiffDbSession {
  private constructor(
    private readonly driver: DriverApplicationTransport,
    private readonly client: InstanceType<TicketDeskModule["TicketDeskClient"]>,
  ) {}

  static async connect(config: DriverIdentityFile): Promise<RiffDbSession> {
    const generated = await loadGenerated();
    const driver = await DriverApplicationTransport.connect({
      socketPath: config.socketPath,
      identity: config.identity,
    });
    const transport = new DriverGeneratedApplicationTransport(driver);
    const client = new generated.TicketDeskClient(transport, 1);
    return new RiffDbSession(driver, client);
  }

  async close(): Promise<void> {
    await this.driver.shutdown();
  }

  async prewarm(organizationId: UuidBytes, ticketId: UuidBytes): Promise<void> {
    await this.pointGetTicket(organizationId, ticketId);
  }

  async pointGetTicket(organizationId: UuidBytes, ticketId: UuidBytes): Promise<boolean> {
    const result = await this.client.getTicket({
      organization_id: formatUuid(organizationId),
      ticket_id: formatUuid(ticketId),
    });
    return result.value.outcome === "Found";
  }

  async pointGetUser(organizationId: UuidBytes, userId: UuidBytes): Promise<boolean> {
    const result = await this.client.getUser({
      organization_id: formatUuid(organizationId),
      user_id: formatUuid(userId),
    });
    return result.value.outcome === "Found";
  }

  async listTicketsByProjectStatus(
    organizationId: UuidBytes,
    projectId: UuidBytes,
    status: string,
    limit: number,
  ): Promise<void> {
    await this.client.listTickets({
      organization_id: formatUuid(organizationId),
      project_id: formatUuid(projectId),
      statuses: [sqlStatusToRiff(status)],
      limit,
    });
  }

  async listOpenTicketsForAssignee(
    organizationId: UuidBytes,
    assigneeId: UuidBytes,
    limit: number,
  ): Promise<void> {
    await this.client.listTicketsByAssignee({
      organization_id: formatUuid(organizationId),
      assignee_id: formatUuid(assigneeId),
      statuses: ["Open"],
      limit,
    });
  }

  async listCommentsForTicket(
    organizationId: UuidBytes,
    ticketId: UuidBytes,
    limit: number,
  ): Promise<void> {
    await this.client.listComments({
      organization_id: formatUuid(organizationId),
      ticket_id: formatUuid(ticketId),
      limit,
    });
  }

  async listProjectMembers(organizationId: UuidBytes, projectId: UuidBytes, _limit: number): Promise<void> {
    await this.client.projectMembers({
      organization_id: formatUuid(organizationId),
      project_id: formatUuid(projectId),
    });
  }

  async ticketDetailPage(organizationId: UuidBytes, ticketId: UuidBytes): Promise<boolean> {
    const result = await this.client.ticketPage({
      organization_id: formatUuid(organizationId),
      ticket_id: formatUuid(ticketId),
    });
    return result.value.outcome === "Found";
  }

  async createComment(comment: CommentSeed): Promise<void> {
    const result = await this.client.createComment({
      body: comment.row.body,
      author_id: formatUuid(comment.row.authorId),
      ticket_id: formatUuid(comment.row.ticketId),
      comment_id: formatUuid(comment.row.commentId),
      idempotency_key: comment.idempotencyKey,
      organization_id: formatUuid(comment.row.organizationId),
    });
    requireOutcome(result.outcome.outcome, "Created");
  }

  async closeTicketWithComment(input: CloseTicketWithCommentSeed): Promise<void> {
    const result = await this.client.closeTicketWithComment({
      body: input.body,
      author_id: formatUuid(input.authorId),
      ticket_id: formatUuid(input.ticketId),
      comment_id: formatUuid(input.commentId),
      idempotency_key: input.idempotencyKey,
      organization_id: formatUuid(input.organizationId),
    });
    requireOutcome(result.outcome.outcome, "Closed");
  }

  async openTicketWithLabels(input: OpenTicketWithLabelsSeed): Promise<void> {
    const result = await this.client.openTicketWithLabels({
      title: input.title,
      label_a: formatUuid(input.labelA),
      label_b: formatUuid(input.labelB),
      ticket_id: formatUuid(input.ticketId),
      project_id: formatUuid(input.projectId),
      assignee_id: formatUuid(input.assigneeId),
      reporter_id: formatUuid(input.reporterId),
      idempotency_key: input.idempotencyKey,
      organization_id: formatUuid(input.organizationId),
    });
    requireOutcome(result.outcome.outcome, "Created");
  }

  async seed(dataset: SeedDataset): Promise<void> {
    // Seed through the generated bounded batch envelope, as the Rust harness
    // does. A sequential `for`/`await` loop ran at single-client rate, which is
    // invisible at `full` scale (19,220 commands) and unusable at `production`
    // scale (~1.1M): a measured run reached 1.4 GB after twenty-one minutes
    // without finishing. Every item remains an ordinary independent generated
    // command with its own durable lifecycle; only the transport is batched.
    const options = { concurrency: SEED_CONCURRENCY };
    const total =
      dataset.organizations.length + dataset.users.length + dataset.projects.length +
      dataset.members.length + dataset.labels.length + dataset.tickets.length +
      dataset.comments.length + dataset.ticketLabels.length;
    let completed = 0;
    const started = Date.now();
    process.stderr.write(
      `riffdb-seed-start\ttotal=${total}\tconcurrency=${SEED_CONCURRENCY}\n`,
    );

    // Phases respect foreign-key order.
    // JSON.stringify is a proxy for the driver's own encoding, not identical to
    // it; the half-budget target above is what absorbs the difference.
    const chunkRowsFor = (inputs: ReadonlyArray<unknown>): number => {
      if (inputs.length === 0) return SEED_MAX_BATCH_ROWS;
      const sample = Buffer.byteLength(JSON.stringify(inputs[0]), "utf8");
      if (sample <= 0) return SEED_MAX_BATCH_ROWS;
      const rows = Math.floor(SEED_FRAME_TARGET_BYTES / sample);
      return Math.max(1, Math.min(SEED_MAX_BATCH_ROWS, rows));
    };

    const runPhase = async <I, O extends { readonly outcome: string }>(
      phase: string,
      inputs: ReadonlyArray<I>,
      call: (
        chunk: ReadonlyArray<I>,
        options: generatedClient.CommandBatchOptions,
      ) => Promise<generatedClient.CommandBatchResult<O>>,
      expected: string,
    ): Promise<void> => {
      const batchRows = chunkRowsFor(inputs);
      for (let offset = 0; offset < inputs.length; offset += batchRows) {
        const chunk = inputs.slice(offset, offset + batchRows);
        const result = await call(chunk, options);
        if (result.items.length !== chunk.length) {
          throw new Error(`seed phase ${phase} returned ${result.items.length} of ${chunk.length}`);
        }
        for (const item of result.items) {
          // A batch reports a per-item failure in `error` rather than rejecting,
          // so a phase that ignored it would seed silently short.
          if (item.error !== undefined) throw item.error;
          if (item.result === undefined) {
            throw new RiffDbLoadError(`seed phase ${phase} item ${item.index} carried no result`);
          }
          requireOutcome(item.result.outcome.outcome, expected);
        }
        completed += chunk.length;
      }
      process.stderr.write(
        `riffdb-seed-progress\tphase=${phase}\tcompleted=${completed}/${total}` +
          `\tbatch_rows=${batchRows}\toverall_ms=${Date.now() - started}\n`,
      );
    };

    await runPhase("organization", dataset.organizations.map(([orgId, name]) => ({
      name, organization_id: formatUuid(orgId),
      idempotency_key: `seed-org-${encodeShort(orgId)}`,
    })), (chunk, o) => this.client.createOrganizationBatch(chunk, o), "Created");

    await runPhase("user", dataset.users.map(([orgId, userId, email, display]) => ({
      email, user_id: formatUuid(userId), display_name: display,
      organization_id: formatUuid(orgId),
      idempotency_key: `seed-user-${encodeShort(userId)}`,
    })), (chunk, o) => this.client.createUserBatch(chunk, o), "Created");

    await runPhase("project", dataset.projects.map(([orgId, projectId, name]) => ({
      name, project_id: formatUuid(projectId), organization_id: formatUuid(orgId),
      idempotency_key: `seed-project-${encodeShort(projectId)}`,
    })), (chunk, o) => this.client.createProjectBatch(chunk, o), "Created");

    await runPhase("member", dataset.members.map(([orgId, projectId, userId, role]) => ({
      role, user_id: formatUuid(userId), project_id: formatUuid(projectId),
      organization_id: formatUuid(orgId),
      idempotency_key: `seed-member-${encodeShort(projectId)}-${encodeShort(userId)}`,
    })), (chunk, o) => this.client.addProjectMemberBatch(chunk, o), "Created");

    await runPhase("label", dataset.labels.map(([orgId, labelId, name]) => ({
      name, label_id: formatUuid(labelId), organization_id: formatUuid(orgId),
      idempotency_key: `seed-label-${encodeShort(labelId)}`,
    })), (chunk, o) => this.client.createLabelBatch(chunk, o), "Created");

    await runPhase("ticket", dataset.tickets.map((ticket) => ({
      title: ticket.title, status: sqlStatusToRiff(ticket.status),
      ticket_id: formatUuid(ticket.ticketId),
      project_id: formatUuid(ticket.projectId),
      reporter_id: formatUuid(ticket.reporterId),
      assignee_id: formatUuid(ticket.assigneeId),
      organization_id: formatUuid(ticket.organizationId),
      idempotency_key: `seed-ticket-${encodeShort(ticket.ticketId)}`,
    })), (chunk, o) => this.client.createTicketBatch(chunk, o), "Created");

    await runPhase("comment", dataset.comments.map((comment) => ({
      body: comment.body, comment_id: formatUuid(comment.commentId),
      ticket_id: formatUuid(comment.ticketId),
      author_id: formatUuid(comment.authorId),
      organization_id: formatUuid(comment.organizationId),
      idempotency_key: `seed-comment-${encodeShort(comment.commentId)}`,
    })), (chunk, o) => this.client.createCommentBatch(chunk, o), "Created");

    await runPhase("ticket_label", dataset.ticketLabels.map(([orgId, ticketId, labelId]) => ({
      label_id: formatUuid(labelId), ticket_id: formatUuid(ticketId),
      organization_id: formatUuid(orgId),
      idempotency_key: `seed-link-${encodeShort(ticketId)}-${encodeShort(labelId)}`,
    })), (chunk, o) => this.client.attachLabelBatch(chunk, o), "Created");

    process.stderr.write(
      `riffdb-seed-progress\tphase=done\tcompleted=${completed}/${total}` +
        `\toverall_ms=${Date.now() - started}\n`,
    );
  }
}

export class RiffDbDriver {
  readonly backendId = RIFFDB_BACKEND_ID;

  constructor(private readonly identity: DriverIdentityFile) {}

  async seed(dataset: SeedDataset): Promise<void> {
    const session = await RiffDbSession.connect(this.identity);
    try {
      await session.seed(dataset);
    } finally {
      await session.close();
    }
  }

  async openSession(): Promise<RiffDbSession> {
    return RiffDbSession.connect(this.identity);
  }
}
